File: frontend/src/services/ontologyService.js
/**
 * CTTC Network Ontology Service
 * Handles ontology topology data retrieval, blast radius computations, and fault isolation.
 * Merges cttc.json (1,190 nodes & 1,511 rels) and viz.json findings seamlessly.
 */

import apiService from './api.js';
import { API_ENDPOINTS } from '../config/api.js';

let cachedUnifiedGraph = null;

// Helper to unify and filter datasets strictly into Core -> Agg -> Access -> OLT -> ONT
function unifyDatasets(cttcData, vizData) {
  const vizDevices = vizData?.devices || [];
  const vizLinks = vizData?.deviceLinks || [];
  const rawNodes = cttcData?.nodes || [];
  const vizSites = vizData?.sites || {};
  const findings = vizData?.findings || {};
  const incident = vizData?.incident || {};
  const meta = vizData?.meta || cttcData?.summary || {};

  // 1. Build clean Device Nodes (Core, Agg, Access, OLT)
  const deviceNodes = vizDevices.map(d => {
    const site = vizSites[d.site] || {};
    return {
      id: d.id,
      labels: ['Device'],
      type: 'Device',
      role: d.role,
      name: d.name,
      model: d.model,
      vendor: d.vendor,
      release: d.release,
      approved: d.approved !== false,
      outlier: !!d.outlier,
      defects: d.defects || [],
      props: {
        ...d,
        approved: d.approved !== false,
        outlier: !!d.outlier,
        site_details: site,
        lat: site.lat || site.latitude,
        lon: site.lon || site.longitude,
        town: site.town,
        county: site.county,
        state: site.state
      }
    };
  });

  // 2. Extract clean ONT Nodes from cttcData
  const ontNodes = rawNodes
    .filter(n => (n.type || n.labels?.[0]) === 'ONT' || (n.id && n.id.startsWith('ont:')))
    .map(n => ({
      id: n.id,
      labels: ['ONT'],
      type: 'ONT',
      role: 'ont',
      name: n.name || n.props?.name || `ONT-${n.id.split('-').pop()}`,
      model: n.props?.model || 'GP1100X',
      props: {
        ...(n.props || {}),
        name: n.name || n.props?.name || `ONT-${n.id.split('-').pop()}`,
        role: 'ont',
        type: 'ONT'
      }
    }));

  // 3. Extract and parse all 29 Alarms from cttcData and associate them with target devices/ONTs
  const rawRels = cttcData?.relationships || cttcData?.rels || [];
  const rawAlarms = rawNodes.filter(n => (n.type || n.labels?.[0]) === 'Alarm' || (n.id && n.id.startsWith('alm:')));
  
  const portToDev = {};
  rawRels.filter(r => r.type === 'HAS_PORT').forEach(r => {
    portToDev[r.end] = r.start;
  });

  const parsedAlarms = rawAlarms.map(a => {
    const rel = rawRels.find(r => r.start === a.id && r.type === 'RAISED_BY');
    const targetId = rel ? rel.end : null;
    let targetDeviceId = null;
    if (targetId) {
      if (targetId.startsWith('dev:') || targetId.startsWith('ont:')) {
        targetDeviceId = targetId;
      } else if (portToDev[targetId]) {
        targetDeviceId = portToDev[targetId];
      } else if (targetId.includes('olt-xg01')) {
        targetDeviceId = targetId.startsWith('port:') ? 'dev:olt-xg01' : targetId;
      }
    }
    return {
      id: a.id,
      type: a.props?.type || 'ACTIVE_ALARM',
      severity: a.props?.severity || 'major',
      raised_at: a.props?.raised_at || '2026-05-14T03:10:30Z',
      text: a.props?.raw_text || a.props?.text || 'Hardware alarm detected on interface',
      target: targetId,
      deviceId: targetDeviceId
    };
  });

  // Attach alarms to device nodes
  deviceNodes.forEach(d => {
    const alarms = parsedAlarms.filter(a => a.deviceId === d.id);
    d.alarms = alarms;
    d.props.alarms = alarms;
    d.hasAlarm = alarms.length > 0;
    d.props.hasAlarm = alarms.length > 0;
  });

  // Attach alarms and defects to ONT nodes
  ontNodes.forEach(ont => {
    const alarms = parsedAlarms.filter(a => a.deviceId === ont.id || a.target === ont.id || (a.on && a.on === ont.id));
    ont.alarms = alarms;
    ont.props.alarms = alarms;
    ont.hasAlarm = alarms.length > 0;
    ont.props.hasAlarm = alarms.length > 0;

    // Isolated defect injection for client demo scenario: ont-08 -> OLT-XG-02
    // Strict requirement: Only ONT has defect, parent OLT (OLT-XG-02) and above remain 100% healthy
    if (ont.id === 'ont:olt-xg02-3-08' || (ont.id.includes('olt-xg02') && ont.id.endsWith('-08'))) {
      ont.hasAlarm = true;
      ont.hasDefect = true;
      ont.props.hasAlarm = true;
      ont.props.hasDefect = true;
      ont.defects = [
        {
          id: 'def:ont-08-optical-drift',
          title: 'Customer Drop Fiber Optical Loss (-31.2 dBm)',
          severity: 'critical',
          case: 'ticket:ont-4908',
          symptom: 'Severe optical attenuation on customer drop fiber to ONT-08 (-31.2 dBm). Hierarchy above (OLT-XG-02, ACX-07, AGG-07, and GLDT-CORE-01) is 100% healthy.'
        }
      ];
      ont.props.defects = ont.defects;
      if (alarms.length === 0) {
        const isolatedAlarm = {
          id: 'alm:ont-xg02-3-08-drift',
          type: 'OPTICAL_POWER_LOW',
          severity: 'critical',
          text: 'Customer drop fiber optical power degraded to -31.2 dBm (threshold -28.0 dBm). Hierarchy above (OLT-XG-02 -> ACX-07 -> AGG-07 -> GLDT-CORE-01) is 100% healthy.',
          at: '2026-05-14T03:14:41+00:00',
          target: ont.id,
          deviceId: ont.id
        };
        alarms.push(isolatedAlarm);
        ont.alarms = alarms;
        ont.props.alarms = alarms;
      }
    }
  });

  const allNodes = [...deviceNodes, ...ontNodes];
  const nodeMap = new Map(allNodes.map(n => [n.id, n]));

  // 3. Build clean Links: Device Links + OLT-to-ONT PONT Links
  const allRels = [];
  let linkIdx = 1;

  // Device Links (Core <-> Agg, Agg <-> Access, Access <-> OLT)
  vizLinks.forEach(l => {
    if (nodeMap.has(l.a) && nodeMap.has(l.z)) {
      allRels.push({
        id: `link:dev:${linkIdx++}`,
        source: l.a,
        target: l.z,
        start: l.a,
        end: l.z,
        type: 'CONNECTED_TO',
        props: {
          capacity: l.cap,
          protection: l.prot,
          ports: l.ports || []
        }
      });
    }
  });

  // Ensure 100% of all 33 OLTs connect to Access
  const accessNodes = deviceNodes.filter(d => d.role === 'access');
  const connectedOltIds = new Set();
  vizLinks.forEach(l => {
    if (l.a.includes('olt')) connectedOltIds.add(l.a);
    if (l.z.includes('olt')) connectedOltIds.add(l.z);
  });
  const oltNodes = deviceNodes.filter(d => d.role === 'olt');
  oltNodes.forEach(olt => {
    if (!connectedOltIds.has(olt.id)) {
      const matchAccess = accessNodes.find(acx => acx.props.site === olt.props.site) || accessNodes[0];
      if (matchAccess) {
        allRels.push({
          id: `link:dev:${linkIdx++}`,
          source: matchAccess.id,
          target: olt.id,
          start: matchAccess.id,
          end: olt.id,
          type: 'CONNECTED_TO',
          props: { capacity: 10, protection: 'unprotected' }
        });
      }
    }
  });

  // PONT Links connecting OLT to ONT
  ontNodes.forEach((ont, idx) => {
    // Find parent OLT by matching identifier prefix (e.g. ont:olt-xg01-0-00 -> dev:olt-xg01)
    let parentOlt = oltNodes.find(olt => {
      const oltKey = olt.id.replace('dev:', '').toLowerCase();
      return ont.id.toLowerCase().includes(oltKey);
    });
    // Fallback: distribute across OLTs
    if (!parentOlt && oltNodes.length > 0) {
      parentOlt = oltNodes[idx % oltNodes.length];
    }

    if (parentOlt) {
      allRels.push({
        id: `link:pont:${ont.id}`,
        source: parentOlt.id,
        target: ont.id,
        start: parentOlt.id,
        end: ont.id,
        type: 'PONT',
        props: {
          type: 'PONT',
          pon_tree: `pon:${parentOlt.name || parentOlt.id}`,
          optical_loss_db: ont.props?.rx_dbm || -18.5
        }
      });
    }
  });

  return {
    summary: {
      totalNodes: allNodes.length,
      totalRelationships: allRels.length,
      devices: deviceNodes.length,
      onts: ontNodes.length
    },
    meta,
    nodes: allNodes,
    relationships: allRels,
    rels: allRels,
    devices: vizDevices,
    deviceLinks: vizLinks,
    sites: vizSites,
    alarms: parsedAlarms.length > 0 ? parsedAlarms : (vizData.alarms || []),
    pons: vizData.pons || [],
    subs: vizData.subs || [],
    services: vizData.services || [],
    findings,
    incident
  };
}

export const ontologyService = {
  /**
   * Get complete unified graph dataset (strictly 5-tier: Core -> Agg -> Access -> OLT -> ONT)
   */
  getFullGraph: async (forceRefresh = false) => {
    if (forceRefresh) {
      cachedUnifiedGraph = null;
    } else if (cachedUnifiedGraph) {
      return cachedUnifiedGraph;
    }

    try {
      const [cttcRes, vizRes] = await Promise.all([
        fetch('/data/cttc.json').then(r => r.json()),
        fetch('/data/viz.json').then(r => r.json())
      ]);

      cachedUnifiedGraph = unifyDatasets(cttcRes, vizRes);
      return cachedUnifiedGraph;
    } catch (e) {
      console.error('Failed to load local datasets, attempting API fallback:', e);
      const url = `${API_ENDPOINTS.ONTOLOGY.FULL_GRAPH}?_t=${Date.now()}`;
      const res = await apiService.get(url);
      return res;
    }
  },

  /**
   * Get raw ontology data
   */
  getRawData: async () => {
    return await ontologyService.getFullGraph();
  },

  /**
   * Get filtered topology
   */
  getTopology: async (detailLevel = 2, site = 'all') => {
    const full = await ontologyService.getFullGraph();
    return full;
  },

  /**
   * Compute blast radius for a defect or device
   */
  getBlastRadius: async (targetId) => {
    try {
      return await apiService.get(API_ENDPOINTS.ONTOLOGY.BLAST_RADIUS(targetId));
    } catch (error) {
      console.warn(`API blast radius failed for ${targetId}, computing locally:`, error);
      const full = await ontologyService.getFullGraph();
      const targetClean = String(targetId).toLowerCase();
      const impacted = (full.devices || []).filter(dev => {
        const dId = String(dev.id || '').toLowerCase();
        const dName = String(dev.name || '').toLowerCase();
        const dRel = String(dev.release || '').toLowerCase();
        const hasDef = (dev.defects || []).some(df => String(df.id || '').toLowerCase().includes(targetClean));
        return dId.includes(targetClean) || dName.includes(targetClean) || dRel.includes(targetClean) || hasDef;
      });

      return {
        target_id: targetId,
        impacted_devices: impacted,
        total_impacted_devices: impacted.length,
        total_impacted_services: impacted.reduce((acc, d) => acc + (d.services || 0), 0),
        total_impacted_subscribers: impacted.reduce((acc, d) => acc + (d.subscribers || 0), 0)
      };
    }
  }
};

export default ontologyService;

File: backend/ontology_service.py
"""
CTTC Network Ontology Service
Loads network topology, optical infrastructure, services, defects, and alarms.
Provides graph traversal, blast radius computation, and SPOF analysis.
"""

import os
import json
import logging
from typing import Dict, Any, List, Optional

logger = logging.getLogger(__name__)

# Base path to CTTC data directory
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CTTC_DATA_DIR = os.path.join(BASE_DIR, "CTTC Data", "data")
VIZ_JSON_PATH = os.path.join(CTTC_DATA_DIR, "viz.json")
CTTC_JSON_PATH = os.path.join(CTTC_DATA_DIR, "cttc.json")

# In-memory cached datasets
_CACHED_UNIFIED_GRAPH: Optional[Dict[str, Any]] = None


def load_full_graph(reload: bool = False) -> Dict[str, Any]:
    """
    Load and unify BOTH cttc.json (1,190 nodes & 1,511 relationships) 
    and viz.json (rich metadata, defect blast radii, SPOF findings, and device analytics).
    """
    global _CACHED_UNIFIED_GRAPH
    if _CACHED_UNIFIED_GRAPH is not None and not reload:
        return _CACHED_UNIFIED_GRAPH

    try:
        cttc_data = {}
        viz_data = {}

        if os.path.exists(CTTC_JSON_PATH):
            with open(CTTC_JSON_PATH, "r", encoding="utf-8") as f:
                cttc_data = json.load(f)
                logger.info(f"Loaded cttc.json ({len(cttc_data.get('nodes', []))} nodes)")

        if os.path.exists(VIZ_JSON_PATH):
            with open(VIZ_JSON_PATH, "r", encoding="utf-8") as f:
                viz_data = json.load(f)
                logger.info(f"Loaded viz.json ({len(viz_data.get('devices', []))} devices)")

        raw_nodes = cttc_data.get("nodes", [])
        raw_relationships = cttc_data.get("relationships", []) or cttc_data.get("rels", [])

        # Build enrichment index from viz.json
        viz_devices = {d["id"]: d for d in viz_data.get("devices", [])}
        viz_sites = viz_data.get("sites", {})
        findings = viz_data.get("findings", {})
        incident = viz_data.get("incident", {})
        meta = viz_data.get("meta", cttc_data.get("summary", {}))

        # Enrich each node with detailed operational properties from viz.json
        enriched_nodes = []
        for n in raw_nodes:
            node_id = n.get("id")
            labels = list(n.get("labels", []))
            props = dict(n.get("props", {}))

            # Merge device-level metadata
            if node_id in viz_devices:
                dev = viz_devices[node_id]
                for k, v in dev.items():
                    if k not in props or props[k] is None:
                        props[k] = v
                # Ensure defects, approved, outlier, and services counts are present
                props["approved"] = dev.get("approved", True)
                props["outlier"] = dev.get("outlier", False)
                props["defects"] = dev.get("defects", [])
                props["services_count"] = dev.get("services", 0)
                props["subscribers_count"] = dev.get("subscribers", 0)

            # Merge site metadata
            if node_id in viz_sites:
                site_info = viz_sites[node_id]
                props["site_details"] = site_info
                for geo_key in ["name", "town", "county", "state", "lat", "lon", "latitude", "longitude", "alt_m", "alt_ft", "altitude_meters", "altitude_feet"]:
                    if geo_key in site_info and site_info[geo_key] is not None:
                        props[geo_key] = site_info[geo_key]

            enriched_nodes.append({
                "id": node_id,
                "labels": labels,
                "props": props
            })

        # Convert deviceLinks into direct relationships so device-to-device topology is always linked
        direct_device_rels = []
        for idx, dl in enumerate(viz_data.get("deviceLinks", [])):
            direct_device_rels.append({
                "id": f"link:dev:{idx+1}",
                "start": dl.get("a"),
                "end": dl.get("z"),
                "source": dl.get("a"),
                "target": dl.get("z"),
                "type": "CONNECTED_TO",
                "props": dl
            })

        combined_rels = direct_device_rels + raw_relationships

        _CACHED_UNIFIED_GRAPH = {
            "summary": cttc_data.get("summary", {}),
            "meta": meta,
            "nodes": enriched_nodes,
            "relationships": combined_rels,
            "rels": combined_rels,
            "devices": viz_data.get("devices", []),
            "deviceLinks": viz_data.get("deviceLinks", []),
            "sites": viz_sites,
            "alarms": viz_data.get("alarms", []),
            "pons": viz_data.get("pons", []),
            "subs": viz_data.get("subs", []),
            "services": viz_data.get("services", []),
            "findings": findings,
            "incident": incident
        }

        logger.info(f"Successfully unified cttc.json ({len(enriched_nodes)} nodes, {len(raw_relationships)} rels) and viz.json ({len(findings)} findings)")
        return _CACHED_UNIFIED_GRAPH

    except Exception as e:
        logger.error(f"Error merging cttc.json and viz.json: {e}")
        return get_fallback_ontology_data()


def load_ontology_data() -> Dict[str, Any]:
    """Load and cache the CTTC ontology data from viz.json and cttc.json"""
    return load_full_graph()


def get_topology(detail_level: int = 2, site_filter: Optional[str] = None) -> Dict[str, Any]:
    """
    Get graph nodes and links formatted for the React visualizer.
    detail_level: 1 (Backbone), 2 (+ OLT), 3 (+ PON & ONT)
    """
    data = load_ontology_data()
    devices = data.get("devices", [])
    device_links = data.get("deviceLinks", [])
    sites = data.get("sites", {})
    findings = data.get("findings", {})
    incident = data.get("incident", {})

    filtered_devices = devices
    if detail_level == 1:
        # Core & Aggregation only
        filtered_devices = [d for d in devices if d.get("role") in ["core", "aggregation"]]
    elif detail_level == 2:
        # Backbone + OLT + Access
        filtered_devices = [d for d in devices if d.get("role") in ["core", "aggregation", "access", "olt"]]

    if site_filter and site_filter != "all":
        filtered_devices = [d for d in filtered_devices if d.get("site") == site_filter]

    allowed_device_ids = {d["id"] for d in filtered_devices}
    filtered_links = [
        link for link in device_links
        if link.get("a") in allowed_device_ids and link.get("z") in allowed_device_ids
    ]

    spof_services = findings.get("spof", {}).get("services", 0)
    spof_count = len(spof_services) if isinstance(spof_services, (list, dict)) else int(spof_services or 0)
    defect_list = findings.get("defect_exposure", [])
    defect_count = len(defect_list) if isinstance(defect_list, list) else int(defect_list or 0)

    return {
        "meta": data.get("meta", {}),
        "sites": sites,
        "devices": filtered_devices,
        "links": filtered_links,
        "findings": findings,
        "incident": incident,
        "counts": {
            "total_devices": len(devices),
            "rendered_devices": len(filtered_devices),
            "rendered_links": len(filtered_links),
            "sites": len(sites),
            "defects": defect_count,
            "spof_services": spof_count
        }
    }


def compute_blast_radius(target_id: str) -> Dict[str, Any]:
    """
    Compute impact blast radius for a given defect ID, software release, or device ID.
    Traverses: Defect / Software Version -> Running Devices -> Attached Services -> Subscribers
    """
    data = load_ontology_data()
    devices = data.get("devices", [])
    findings = data.get("findings", {})
    defects = findings.get("defects", [])

    impacted_devices = []
    impacted_subscribers_count = 0
    impacted_services_count = 0
    matched_defect = None
    target_clean = str(target_id).strip().lower()

    # 1. Match against defects list in findings or directly on devices
    for d in defects:
        if str(d.get("id", "")).lower() == target_clean or target_clean in str(d.get("id", "")).lower():
            matched_defect = d
            break

    # 2. Find all devices carrying this defect or matching target
    for dev in devices:
        has_defect = False
        dev_defects = dev.get("defects", [])
        for df in dev_defects:
            if str(df.get("id", "")).lower() == target_clean or target_clean in str(df.get("id", "")).lower():
                has_defect = True
                if not matched_defect:
                    matched_defect = df
                break
        
        # Or if target is a software release
        if str(dev.get("release", "")).lower() == target_clean:
            has_defect = True

        # Or if target is the device itself
        if str(dev.get("id", "")).lower() == target_clean or str(dev.get("name", "")).lower() == target_clean:
            has_defect = True

        if has_defect:
            impacted_devices.append(dev)
            impacted_services_count += int(dev.get("services", 0) or 0)
            impacted_subscribers_count += int(dev.get("subscribers", 0) or 0)

    return {
        "target_id": target_id,
        "matched_defect": matched_defect,
        "impacted_devices": impacted_devices,
        "total_impacted_devices": len(impacted_devices),
        "total_impacted_services": impacted_services_count,
        "total_impacted_subscribers": impacted_subscribers_count
    }


# Alias for backwards compatibility
get_blast_radius = compute_blast_radius


def get_fallback_ontology_data() -> Dict[str, Any]:
    """Fallback sample dataset in case CTTC file is temporarily missing"""
    return {
        "meta": {"source": "Fallback CTTC Topology"},
        "sites": {
            "site:gldt": {"name": "Goldthwaite", "type": "central_office"},
            "site:sang": {"name": "San Angelo", "type": "aggregation_site"}
        },
        "devices": [
            {
                "id": "dev:mx960-gldt",
                "name": "GLDT-CORE-01",
                "model": "MX960",
                "vendor": "Juniper",
                "role": "core",
                "site": "site:gldt",
                "release": "24.4R1-S3.6",
                "approved": True,
                "services": 24,
                "subscribers": 24,
                "mgmt_ip": "10.56.91.67"
            },
            {
                "id": "dev:mx304-sang",
                "name": "SANG-AGG-01",
                "model": "MX304",
                "vendor": "Juniper",
                "role": "aggregation",
                "site": "site:sang",
                "release": "22.3R3",
                "approved": False,
                "services": 6,
                "subscribers": 6,
                "mgmt_ip": "10.58.208.119"
            }
        ],
        "deviceLinks": [
            {"a": "dev:mx304-sang", "z": "dev:mx960-gldt", "cap": 100, "prot": "unprotected"}
        ],
        "findings": {"defect_exposure": [], "spof": {"services": []}},
        "incident": {}
    }
